### PR TITLE
refactor: Implement `extras` attribute for Plugin

### PR DESCRIPTION
**Pull request by elenakrittik**
This PR implements the `PluginMetadata.extras` attribute, which supersedes `PluginMetadata.category`. This allows more data than just category to be easily stored on a per-plugin basis.

Any usage of `Plugin.category` or `PluginMetadata.category` will not result in `DeprecationWarning`s. Several `.. deprecated::` directives were also added to docstrings.

Also added a `Plugin.extras` shorthand property.

### CODE REVIEW

File: examples/main.py
Comment (Sharp-Eyes) [inline examples/main.py:1]: Please undo this, this is out of scope for this PR and the import is intentionally down there for a more natural reading flow.

File: src/disnake/ext/plugins/plugin.py [hunk 1]
```diff
@@ -172,11 +171,14 @@ class Plugin(t.Generic[BotT]):
     logger: Optional[Union[:class:`logging.Logger`, :class:`str`]]
         The logger or its name to use when logging plugin events.
         If not specified, defaults to `disnake.ext.plugins.plugin`.
+    extras: Dict[:class:`str`, Any]
```
Comment (Sharp-Eyes) [inline src/disnake/ext/plugins/plugin.py]: This doesn't quite check out-- `Parameters` are the `__init__` parameters, and `extras` is not one of them as you used `**kwargs` there.

I think it'd be most clear if we used `**extras` both in the docstring here and in the method signature?

File: src/disnake/ext/plugins/plugin.py [hunk 2]
```diff
@@ -314,11 +320,6 @@ def name(self) -> str:
         """The name of this plugin."""
         return self.metadata.name
 
-    @property
-    def category(self) -> t.Optional[str]:
-        """The category this plugin belongs to."""
-        return self.metadata.category
-
```
Comment (Sharp-Eyes) [inline src/disnake/ext/plugins/plugin.py]: (This is not necessarily a change I **want** to see changed, I'm mostly just giving an idea-- what do you think?)

Maybe it's an idea to leave this property in for backward compatibility and instead just have it `return self.extras.get("category")`?

**Review by Sharp-Eyes — CHANGES_REQUESTED** [review]
Thanks for the PR!

There's a couple small things I'd like to see changed, but outside of that it's mostly good.

File: src/disnake/ext/plugins/plugin.py (hunk 2 above)
Comment (elenakrittik) [reply src/disnake/ext/plugins/plugin.py]: Maybe it's best then to also turn PluginMetadata.category into a property and emit a DeprecationWarning there, also we might retain all the category params for now but set category to extras instead of metadata and emit a DeprecationWarning

File: src/disnake/ext/plugins/plugin.py (hunk 1 above)
Comment (elenakrittik) [reply src/disnake/ext/plugins/plugin.py]: Good catch, thanks! Changed in 33ac4987086a8bc07e8877b0455c557c0aedbb0b

File: examples/main.py
Comment (elenakrittik) [reply examples/main.py:1]: Sorry, pre-commit for some reason ran black and isort against example folder too. Reverted in 33ac4987086a8bc07e8877b0455c557c0aedbb0b

File: src/disnake/ext/plugins/plugin.py (hunk 2 above)
Comment (Sharp-Eyes) [reply src/disnake/ext/plugins/plugin.py]: > also we might retain all the category params for now but set category to extras instead of metadata and emit a DeprecationWarning

Yeah, that's pretty much what I was thinking. Existing code using a category kwarg would keep working as it'd be picked up and stored in extras anyways, so updating the property to get it from extras and optionally adding a deprecation warning should do the trick.

Comment (elenakrittik) [reply src/disnake/ext/plugins/plugin.py]: Main changes in 0b1b65b42fd7f2e6c2eb364681c1e7c02c80224f, and also documented PluginMetadata in 88bd109b7ca3804150c56c5ffc1881a26ecb25f7. Since there's quite a lot of unusual changes, i tried explaining them in the updated original PR message.

**Comment by Sharp-Eyes** [comment]
(Just so we can find this in the future if the need ever arises)

After some [discussion in the Disnake discord](https://discord.com/channels/808030843078836254/1075807043803426886/1118488482315243580), we came up with a better solution.

**Comment by elenakrittik** [comment]
Done!

File: src/disnake/ext/plugins/plugin.py [hunk 3]
```diff
@@ -258,16 +321,25 @@ def __init__(
...
         )
 
+        if category is not None:
+            warnings.warn(
+                "Passing `category` to `Plugin` is deprecated. Use `extras` instead.",
+                DeprecationWarning,
+            )
+
+            self.metadata.extras["category"] = category
+
```
Comment (Sharp-Eyes) [inline src/disnake/ext/plugins/plugin.py]: This is pretty pointless. After deprecation of the category keyword-arg, it'll automatically be picked up by **extras anyways.

File: src/disnake/ext/plugins/plugin.py [hunk 4]
```diff
@@ -94,13 +95,68 @@ class SlashCommandParams(AppCommandParams, total=False):
 
 @dataclasses.dataclass
 class PluginMetadata:
+    """Represents metadata for a :class:`Plugin`
```
Comment (Sharp-Eyes) [inline src/disnake/ext/plugins/plugin.py]: Very minor nitpick-- missing period at the end of the sentence here.

**Review by Sharp-Eyes — CHANGES_REQUESTED** [review]
Two minor points, the rest is looking good ^^

File: src/disnake/ext/plugins/plugin.py (hunk 3 above)
Comment (elenakrittik) [reply src/disnake/ext/plugins/plugin.py]: Clever point, thanks! 451310ccbc09b1dcf20091cef3268523f47f6e51

File: src/disnake/ext/plugins/plugin.py (hunk 4 above)
Comment (elenakrittik) [reply src/disnake/ext/plugins/plugin.py]: 451310ccbc09b1dcf20091cef3268523f47f6e51

**Review by Sharp-Eyes — APPROVED** [review]
LGTM ^^
